### PR TITLE
[ENH] BEP044 - Stim-BIDS

### LINKED ISSUE #2451: BEP 044: Stimuli Project Tracking

### PR DESCRIPTION
**Pull request by neuromechanist**
This PR addresses #153 by introducing specifications for handling and standardizing stimulus files and their annotations within the BIDS specifications. The changes focus on improving the organization, referencing, and metadata of stimulus files to enhance consistency, reusability, and efficiency.

> [!IMPORTANT]
> HTML render of the stimuli page : https://bids-specification--2022.org.readthedocs.build/en/2022/modality-agnostic-files/stimuli.html

> [!NOTE]
> The issue: #153
> [Google Doc](https://docs.google.com/document/d/1DoQghbJlQksCHEFs0boT816p3ejX2Hd9l-OWDTXFtV8/edit?pli=1&tab=t.0#heading=h.ab3cjzoaom6d) 
> Example PR: https://github.com/bids-standard/bids-examples/pull/433

Known issues:
- The validator fails because extensions in `/src/schema/rules/files/raw` require `datatype`. Stimuli might be a special data type that can only be present at the root of the dataset. So, the datatype field is missing for now.
- There are some style errors by the remark validators

cc: @bids-standard/bep044 and @monique2208

### CODE REVIEW

File: src/modality-specific-files/stimuli.md [hunk 1]
```diff
@@ -1,18 +1,25 @@
 # Stimuli
 
-Stimulus files should be stored in the `/stimuli` directory under the root directory of the dataset. The `/stimuli` directory can contain subdirectories to organize the stimulus files. Stimulus files SHOULD follow the BIDS naming conventions and SHOULD be referenced in the `events.tsv` file using the `stim_id` column. The `stim_id` column represents unique identifiers for the stimuli. Additional information about the stimuli and their annotations can be provided in the `stimuli.tsv`, `stimuli.json`, `annotations.tsv`, `annotations.json`, and `stim-<label>.json` files.
+## Stimulus Files Organization
 
-Standardizing stimulus files and their annotations within the BIDS specifications offers several advantages:
+Stimulus files MUST be stored in the `/stimuli` directory under the root directory of the dataset.
+The `/stimuli` directory can contain subdirectories to organize the stimulus files.
+Stimulus files MUST follow the BIDS naming conventions and SHOULD be referenced in the `events.tsv`
```
Comment (VisLab) [inline src/modality-specific-files/stimuli.md:7]: change "SHOULD be" to "are".

This sounds as though you are expected to reference every stimulus file in the events.tsv column.

File: src/modality-specific-files/stimuli.md [hunk 2]
```diff
@@ -87,11 +64,22 @@ A guide for using macros can be found at
 -->
 {{ MACROS___make_sidecar_table("stimulus.Stimulus") }}
 
-## Stimuli.tsv and Stimuli.json
+In some cases, such as observing the copyright of a stimulus file, the actual stimulus file may not be shared. In such cases, the `stim-<label>_<suffix>.json` file SHOULD be used to provide metadata about the stimulus file, including the license, copyright, URL, and description.
```
Comment (VisLab) [inline src/modality-specific-files/stimuli.md:67]: Up above the underbar was removed before the suffix:  <suffix> not _<suffix>.  It should be consistent.

File: src/modality-specific-files/stimuli.md [hunk 3]
```diff
@@ -87,11 +64,22 @@ A guide for using macros can be found at
...
+    "Copyright": "Lab 2023",
+    "URL": "https://example.com/stimuli/",
+    "Description": "Collection of face images, tones, and movie clips used in the experiment"
+}
+```
 
-### Stimuli.tsv
+## Stimuli Description (`stimuli.tsv`)
+
+The `stimuli.tsv/json` files are used to provide information about the stimuli based on their `stim_id`. This file is similar in usage as `participants.tsv`, `scans.tsv` and `sessions.tsv`, which list descriptions about subjects, scans and sessions, respectively. The `stimluli.tsv/json` files MUST be placed in the `/stimuli` directory.
```
Comment (VisLab) [inline src/modality-specific-files/stimuli.md:82]: The slashes before "json" and "tsv"  make it sound like they should be in subdirectories.  This is confusing.

File: src/modality-specific-files/stimuli.md [hunk 4]
```diff
@@ -103,36 +91,69 @@ and a guide for using macros can be found at
...
+stim-movie01  video    https://example.com/movies/movie01.mp4  n/a         Studio XYZ  A clip from copyrighted movie              false
+```
 
-### Stimuli.json
+The `stimuli.json` file provides detailed descriptions of the columns in the `stimuli.tsv` file. There MAY be extra entries in the `stimuli.json` in addition to the columns in the `stimuli.tsv` to provide more details about the stimulus.
 
-The `stimuli.json` file provides detailed descriptions of the columns in the `stimuli.tsv` file. There can be extra entries in the `stimuli.json` in addition to the columns in the `stimuli.tsv` to provide more details about the stimulus.
+In cases where the stimulus is not shared, the `stimuli.tsv` file can be used to provide metadata about the stimuli, including the license, copyright, URL, and description. This is simialr to the use of `stim-<label>_<suffix>.json` files for individual stimuli files. In the case of conflict between the metadata in the `stimuli.tsv` and `stim-<label>_<suffix>.json` files, the metadata in the `stim-<label>_<suffix>.json` file takes precedence.
```
Comment (VisLab) [inline src/modality-specific-files/stimuli.md:105]: Again,  here _<suffix>.  It should be consistent.  The _<suffix>  appears in several places as does <suffix> without the underbar.  I'm not going to mark them further --- just need to be consistent.

File: src/modality-specific-files/stimuli.md [hunk 5]
```diff
@@ -103,36 +91,69 @@ and a guide for using macros can be found at
...
 
-The `annotations.tsv` and `annotations.json` files are used to provide additional information about the annotations associated with the stimuli. These files should be placed in the `/stimuli` directory.
+| stimulus_id  | type  | description | HED | NSD_id | COCO_id |
+|--------------|-------|-------------|-----|--------|---------|
+| stim-nsd02951 | image | an open market full of people and piles of vegetables | ((Item-count, High), Ingestible-object), (Background-view, ((Human, Body, Agent-trait/Adult), Outdoors, Furnishing, Natural-feature/Sky, Urban, Man-made-object)) | 2951 | 262145 |
 
-### Annotations.tsv
+However, for time-varying stimuli, such as audio or video, it is RECOMMENDED to use specific annotations files in the form of `stim-<label>_annot-<label>_events.tsv/json` to store the annotations. These files have the same structure as the `events.tsv/json` files and are used to store annotations for the stimuli. There can be multiple annotation files for a single stimulus file, each with a unique annotation label. The annotation files MUST be stored in the `/stimuli` directory.
```
Comment (VisLab) [inline src/modality-specific-files/stimuli.md:114]: again the tsv/json is confusing.  I think these should be written out or use some other notation.

**Review by VisLab — APPROVED** [review]

File: src/modality-specific-files/stimuli.md (hunk 5 above)
Comment (neuromechanist) [reply src/modality-specific-files/stimuli.md:114]: Removed the `JSON` mentions as they are dependent to the `TSV` files.

File: src/modality-specific-files/stimuli.md (hunk 2 above)
Comment (neuromechanist) [reply src/modality-specific-files/stimuli.md:67]: Thanks, there was one from the tree with the miggning underbar. Overall, I used either *suffix* or `_<suffix>`, which hopeully have clear differences. Please feel free to modify.

**Comment by Remi-Gau** [comment]
will do a bit of clean up to get less red in CI and maybe see if we can get the HTML version of the BEP to render

**Comment by Remi-Gau** [comment]
HTML: stimuli page
https://bids-specification--2022.org.readthedocs.build/en/2022/modality-specific-files/stimuli.html

File: src/modality-specific-files/stimuli.md [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+{{ MACROS___make_sidecar_table("stimuli.Stimuli") }}
+
+In some cases, such as observing the copyright of a stimulus file, the actual stimulus file may not be shared. In such cases, the `stim-<label>_<suffix>.json` file SHOULD be used to provide metadata about the stimulus file, including the license, copyright, URL, and description.
+
+### Example `stim-<label>_<suffix>.json`
+
+```JSON
+{
+    "License": "CC-BY-4.0",
+    "Copyright": "Lab 2023",
```
Comment (yarikoptic) [inline src/modality-specific-files/stimuli.md:75]: we might want to clarify what to state here and whether it should include year , and overall format

may be we should follow https://reuse.software/tutorial/#step-2 and SPDX (reuse uses it too) for license definitions.

File: src/modality-specific-files/task-events.md [hunk 7]
```diff
@@ -400,3 +400,106 @@ A guide for using macros can be found at
...
+    "stim_id": {
+        "LongName": "Stimulus identifier",
+        "Description": "Represents a unique identifier for the stimulus presented at the given onset time. Links to files and annotations in the /stimuli directory."
+    }
+}
+```
+
+The `stim_id` in the events file links to corresponding files:
+
+```Text
```
Comment (yarikoptic) [inline src/modality-specific-files/task-events.md:481]: needs MACROS___make_filetree_example until someone smart implements parsing... ref: https://github.com/bids-standard/bids-specification/pull/2014#issuecomment-2552062712

File: src/schema/rules/files/raw/audio.yaml [hunk 8]
```diff
@@ -0,0 +1,23 @@
+---
+audio:
+  datatypes:
+    - stimuli
+  suffixes:
+    - audio
+  extensions:
+    - .wav
+    - .WAV
```
Comment (yarikoptic) [inline src/schema/rules/files/raw/audio.yaml:9]: I do not think we should allow/encourage mixed casing allowed!  Let's stick to lower case . we have .json and .tsv and no .JSON and .TSV . I do not see why these files need to be different

File: src/schema/objects/suffixes.yaml [hunk 9]
```diff
@@ -877,3 +877,40 @@ unloc:
   description: |
     MRS acquisitions run without localization.
     This includes signals detected using coil sensitivity only.
+audio:
+  value: audio
+  display_name: Audio stimulus file
+  description: |
+    Any data file used as an audio stimulus. Media files usually have an audio file type (such as wav, mp3, aac, ogg).
+    The JSON file associated with each media file should contain information such as License (RECOMMENDED), Copyright
+    (RECOMMENDED), URL (OPTIONAL), and Description (OPTIONAL) to describe the origin and the nature of the media.
```
Comment (yarikoptic) [inline src/schema/objects/suffixes.yaml:886]: I do not think it is advisable to describe all the fields and whether they are required or not -- this duplicates information formalized elsewhere so would be duplication and thus prone to get out of date etc.  

```suggestion
    The JSON file associated with each media file should contain information to describe the origin and the nature of the media.
```

here and similarly below for others

File: src/modality-specific-files/stimuli.md [hunk 10]
```diff
@@ -0,0 +1,161 @@
...
+    "[stim-<label>[_part-<label>]_<suffix>.<extension>]": "",
+    "[stim-<label>[_part-<label>]_<suffix>.json]": "",
+    "[[stim-<label>_]annotations.tsv]": "",
+    "[[stim-<label>_]annotations.json]": "",
+    "[stim-<label>[_part-<label>]_annot-<label>_events.tsv]": "",
+    "[stim-<label>[_part-<label>]_annot-<label>_events.json]": ""
+    }
+  }) }}
+
+Note: The presence of `stimuli.tsv` file indicates that the content of the `/stimuli` directory follows this BIDS specification for stimulus organization. This structure is planned to become mandatory in BIDS 2.0.
```
Comment (yarikoptic) [inline src/modality-specific-files/stimuli.md:40]: Is this a first time we have something like this?  I believe we do not even have similar conditioning for `derivaties/` yet... I just wonder if validator is "ready" or what needs tobe done?  WDYT @effigies ?

**Review by yarikoptic — DISMISSED** [review]
initial quick pass with little recommendations etc

File: mkdocs.yml [hunk 11]
```diff
@@ -20,6 +20,7 @@ nav:
           - Near-Infrared Spectroscopy: modality-specific-files/near-infrared-spectroscopy.md
           - Motion: modality-specific-files/motion.md
           - Magnetic Resonance Spectroscopy: modality-specific-files/magnetic-resonance-spectroscopy.md
+          - Stimuli: modality-specific-files/stimuli.md
```
Comment (Remi-Gau) [inline mkdocs.yml:23]: I put it there for now as the file is in modality specific folder, but I think this stimuli BEP should be 'modality agnostic".

- See also https://github.com/bids-standard/bids-specification/issues/1751

@oesteban

File: src/schema/objects/entities.yaml [hunk 12]
```diff
@@ -194,20 +194,12 @@ part:
...
-    The possible options are `"rad"` or `"arbitrary"`.
-
-    When there is only a magnitude image of a given type, the `part` entity MAY be
-    omitted.
+    This entity is used to indicate which component of a complex
+    representation is being stored. For MRI data, it indicates which component
+    of the complex signal is represented in voxel data. For stimulus files, it can
+    be used to distinguish different parts of a single stimulus, such as chapters
+    in an audiobook or segments of a long movie (for example, `part-1`, `part-2`,
+    `part-epilog`, `part-chapter1`).
```
Comment (Remi-Gau) [inline src/schema/objects/entities.yaml:202]: we probably do not want to erase all the details from the previous definition

can we have several definition for an entity : mean different things depending on the datatype or suffix... ?

File: tools/schemacode/src/bidsschematools/tests/test_render_text.py
Comment (Remi-Gau) [inline tools/schemacode/src/bidsschematools/tests/test_render_text.py:1]: very hacky ways to make the tests pass

File: mkdocs.yml (hunk 11 above)
Comment (oesteban) [reply mkdocs.yml:23]: I agree @Remi-Gau

Comment (neuromechanist) [reply mkdocs.yml:23]: +1.
IMHO, the same goes for task events.

File: src/modality-specific-files/stimuli.md [hunk 13]
```diff
@@ -0,0 +1,161 @@
...
+    "License": "CC-BY-4.0",
+    "Copyright": "Lab 2023",
+    "URL": "https://example.com/stimuli/",
+    "Description": "Collection of face images, tones, and movie clips used in the experiment"
+}
+```
+
+## Stimuli Description (`stimuli.tsv`)
+
+The `stimuli.tsv` files are used to provide information about the stimuli based on their `stim_id`. This file is similar in usage as `participants.tsv`, `scans.tsv` and `sessions.tsv`, which list descriptions about subjects, scans and sessions, respectively. The `stimluli.tsv` files MUST be placed in the `/stimuli` directory.
```
Comment (dorahermes) [inline src/modality-specific-files/stimuli.md:83]: Typo in line83: stimluli.tsv --> stimuli.tsv

Comment (neuromechanist) [reply src/modality-specific-files/stimuli.md]: Thanks, resolved.

File: src/modality-specific-files/stimuli.md [hunk 14]
```diff
@@ -0,0 +1,161 @@
...
+    "[stim-<label>[_part-<label>]_<suffix>.<extension>]": "",
+    "[stim-<label>[_part-<label>]_<suffix>.json]": "",
+    "[[stim-<label>_]annotations.tsv]": "",
+    "[[stim-<label>_]annotations.json]": "",
+    "[stim-<label>[_part-<label>]_annot-<label>_events.tsv]": "",
+    "[stim-<label>[_part-<label>]_annot-<label>_events.json]": ""
+    }
+  }) }}
+
+Note: The presence of the `stimuli.tsv` file indicates that the content of the `/stimuli` directory follows this BIDS specification for stimulus organization. This structure is planned to become mandatory in BIDS 2.0.
```
Comment (yarikoptic) [inline src/modality-specific-files/stimuli.md:40]:
```suggestion
Note: The presence of the `stimuli.tsv` file indicates that the content of the `/stimuli` directory follows this BIDS specification for stimulus organization.
```

I do not think we have yet specific plan for that specific aspect... If there is one, let's point to the issue.

**Comment by neuromechanist** [comment]
Obserinvg #750 and the [`_stim.<extension>` suffix](https://bids-specification.readthedocs.io/en/stable/glossary.html#stim-suffixes), and a discussion pursued last month in #1771, we should clarify when to use the `stimuli` directory and when to use subject-based `_stim.<extension>` files.
IMO, the difference is clear, stimulus files in the `stim` directory are used for >1 subject while the subject-based `_stim.<extension>` is exclusively for the the subject/session where the file is located.

File: mkdocs.yml (hunk 11 above)
Comment (neuromechanist) [reply mkdocs.yml:23]: Resolving in favor of a potential PR to resolve #1751

File: src/modality-specific-files/stimuli.md [hunk 15]
```diff
@@ -0,0 +1,161 @@
...
+Note: The presence of `stimuli.tsv` file indicates that the content of the `/stimuli` directory follows this BIDS specification for stimulus organization. This structure is planned to become mandatory in BIDS 2.0.
+
+### Stimulus File Formats
+
+The following table lists the supported stimulus file formats and their corresponding suffixes. The suffixes are used to identify the type of stimulus file and are appended to the `stim-<label>` prefix in the file name.
+
+| suffix      | extensions                      | description                  |
+| ----------- | ------------------------------- | ---------------------------- |
+| audio       | `.wav`, `.mp3`, `.aac`, `.ogg`  | Audio-only stimulus files    |
+| image       | `.jpg`, `.png`, `.svg`          | Static visual stimulus files |
```
[the remark this answers is not in the document]
Comment (neuromechanist) [reply src/modality-specific-files/stimuli.md]: Implemented in e7c0739

File: src/modality-specific-files/stimuli.md (hunk 6 above)
Comment (neuromechanist) [reply src/modality-specific-files/stimuli.md]: Addressed in c8ffa63935a6eb0dda4fd735da7b141b9a24ce02

File: src/schema/objects/suffixes.yaml (hunk 9 above)
Comment (neuromechanist) [reply src/schema/objects/suffixes.yaml]: Addressed in 8178a3f98241af1ea53ff145d457ab6508949db3

File: src/schema/objects/entities.yaml (hunk 12 above)
Comment (neuromechanist) [reply src/schema/objects/entities.yaml:202]: Thanks, addressed in eafdac9d9baf494ab8aff89dc404b4c82648aead

**Comment by VisLab** [comment]
I think this extension will be a very valuable addition to BIDS. I have read through the HTML rendered version and have the following typo corrections:

1.  In the Stimuli Description section the `stimuli.tsv` table has a lot of `nan` elements.
2.   In the example for `stimuli.tsv` below this table, the `present` column isn't explained (or aligned).  Maybe it is one of the `nan` values in 1.?
3. In the Annotation Description section, the `annotations.tsv` table also has `nan` entries.
4. In the Referencing Stimulus Identifiers section the `./stimuli` directory  should be in code font.

**Comment by neuromechanist** [comment]
> I think this extension will be a very valuable addition to BIDS. I have read through the HTML rendered version and have the following typo corrections:

Thanks very much @VisLab. The `nan`s were due to the columns not being defined in `columns.yaml` They should be ok now.

**Comment by neuromechanist** [comment]
@rwblair and I just discussed the possibilities and challenges to write schema rules for this PR 😍.
Some rules can be enforced readily, but some other would require more development and reflection.
We will try to come up with a list of the rules and their required implementation effort as a next step.

**Review by dorahermes — APPROVED** [review]
lgtm

File: src/modality-agnostic-files/stimuli.md [hunk 16]
```diff
@@ -0,0 +1,160 @@
...
+The inputs for this macro can be found in the directory
+  src/schema/rules/files/raw
+and a guide for using macros can be found at
+ https://github.com/bids-standard/bids-specification/blob/master/macros_doc.md
+-->
+{{ MACROS___make_root_filename_template(
+   "raw",
+   path="stimuli")
+}}
+
```
Comment (yarikoptic) [inline src/modality-agnostic-files/stimuli.md:34]: here it seems to not render entirely as desired - `-<label>` (and `<index>`? for part) is missing

image

File: src/modality-agnostic-files/events.md [hunk 17]
```diff
@@ -397,3 +397,35 @@ A guide for using macros can be found at
...
 [any TSV file](../common-principles.md#tabular-files) to specify, for
 example, the units of the recorded time series for each column.
+
+## Stimulus Organization
+
+For standardized organization and reuse of stimulus files, events may reference stimuli using two approaches:
+
+### Using `stim_file`
+
+Reference stimulus files directly using the `stim_file` column, where values represent the relative path to the stimulus file within the `/stimuli` directory:
```
Comment (yarikoptic) [inline src/modality-agnostic-files/events.md:423]: @neuromechanist I do not want to derail this BEP, but it might as well be included here (or later).

I would like your feedback on our 
- https://github.com/ReproNim/reprostim/issues/14

use case, where we will have a distinct video record for **every** MRI data file in the dataset (thus per subject/[session/]).  I envisioned placing them nearby the data files:

- https://github.com/bids-standard/bids-specification/pull/750

and that is reflective somehow of @bendichter 's 

- https://github.com/bids-standard/bids-specification/pull/2231

where it is a audiovideo record of behavior, instead of audiovideo record of stimuli.

One logical way could be to allow for `stimuli/sub-<label>/[ses-<label>/]...` hierarchy but that feels suboptimal since if unique per sub/ses -- better just go nearby the data.

I would say, we (with @vmdocua) would just add `stim-` entity to the file, associated with the data file, so e.g. for every `_events.tsv` level file it would produce the one with extra `_stim-reprostim_audiovideo.{mkv,json}` .

So in the long run, I would need to advocate allowing for such files in the hierarchy, and that will use `stim` entity then, with `stimuli.tsv` on top level providing description for what that `reprostim` means.

**Comment by neuromechanist** [comment]
Very nice @yarikoptic. Indeed recording the stimulus/subject individual presentations could unlock accounting for even more variabilities throughout the session.

1. I agree that individual stimuli presentation files should go into the subject/session files

2. This might be a sidetone, but I'd think that #750 and #2231 are basically two different things, and that (some form of) both should exist, somehow. We should identify if the video file is a stimulus or behavior (or both?) (through sidecar metadata?).

Hopefully we should not overuse `stim` (especially if the recording captures both subject and stimulus, then it is not stimulus) and we could use something more common like `recording`: 

`sub-01_task-forrest_split-01_recording-reprostim_audiovideo.mp4`
Ideally, we should allow:
`sub-01_task-forrest_split-01_recording-reprostim_events.tsv`
as the `audiovideo.mp4` is a data file like any other `bold` or `eeg` data file that would deserve an `events.tsv` file ;D.

But, in the case that only the presented stimuli is recorded, like presenting randomized NSD data, capturing the video feed and sharing the presented video per participant (to make sure that event times are correct captured?):
`sub-01_task-nsd_stim.mp4` would be enough
and then we have the usual `sub-01_task-nsd_events.tsv` to match that, as stimulus file is "assumed" to accompany a data file (Bold or EEG for that matter).

I hope that I could convey this not super twisted. LMK.

Either way, both should not have a bearing on BEP044. One of the reasons not to push BEP044 out now is to make sure that the use-cases are mature enough and that tools can sufficiently use it. Hence, working on https://github.com/Annotation-Garden/HEDit. So, please bring on more use cases :D.

**Comment by yarikoptic** [comment]
The summary of my "mental" excercise detailed in
- https://github.com/bids-standard/bids-2-devel/issues/97

TL;DR summary for this, I think I will go with @vmdocua for 

    sub-01_task-forrest_stim-reprostim_audiovideo.mp4

ATM, although a potentially more kosher and permitting for mix-in with e.g. `{datamodality}/` folder could be 

    sub-01_task-forrest_mod-stimuli_recording-reprostim_audiovideo.mp4

thus allowing for smth like following to be nearby

    sub-01_task-forrest_mod-beh_recording-wallcamera_audiovideo.mp4

thus clearly separating and annotating the "datatype" (in _mod-) to tell `stim` and `beh` apart.

<details><summary><h3>Little more detailed mental walk through</h3></summary>

I think has good generalization principle I filed in https://github.com/bids-standard/bids-2-devel/issues/97 for this PR would sound like:

- [stim-...] entity, and thus having `stimuli.{tsv,gz}` should remain **optional** under `stimuli/` as in the cases where there was a single stimuli or there are other entities (e.g. `_recording`, `_annot`) are present to tell stimuli files apart. So kinda "better compatibility" with what we have now already (cons: harder to determine if to validate file names under `stimuli/`)
- files with stimuli under specific `sub/[ses/]datatype` folders could gain either
   - `_mod-stim` entity to tell them apart from other files in those folders (e.g. recordings of behavior under `beh/` folders from bep047) or just to announce their "function-in-life" as the stimuli files, or
   - `_stim-<label>` entities if multiple `stim-`uluses for the same data recording.
- generally, we could have define `stim/` folder at `sub/[ses/]` level, to follow the `beh/` and other datatypes, to group stimuli per sub/ses, and then avoid having `_mod-stim` under those since implied, if so was desired  but I think it is not needed hence

Then as a result, for my use-case, since I have only 1 type of stimulus recording, if I were to place stimuli files near data file e.g. `sub-01_task-forrest_bold.nii.gz` , I think I should name it in shortest form as

    sub-01_task-forrest_mod-stim_audiovideo.mp4

and following #2283 we could have `modalities.tsv` (not to be done in this BEP) which could then be used to describe characteristics of `stim` here.  That would also allow to e.g. carry nearby behavior recording (e.g. video camera of the body in the magnet to e.g. assess movements!) under   `sub-01_task-forrest_mod-beh_audiovideo.mp4`.

or if I wanted to be descriptive for my case, although not needed to tell files apart, could still have either 

    sub-01_task-forrest_mod-stim_recording-reprostim_audiovideo.mp4

or the "mixin" between the two as using the entity with a label would define the data type (stimuli) and provide characteristic:

    sub-01_task-forrest_stim-reprostim_audiovideo.mp4

where in stimuli.tsv I would describe `reprostim` value.

</details>

File: src/modality-agnostic-files/stimuli.md (hunk 16 above)
Comment (yarikoptic) [reply src/modality-agnostic-files/stimuli.md:34]: also note that there is no `.json` sidecar mentioned for all those `.<extension>` files!  "explicit better than implicit" and for consistency might be worth doing like done for BEP047 https://bids-specification--2231.org.readthedocs.build/en/2231/modality-specific-files/behavioral-experiments.html

**Comment by yarikoptic** [comment]
@neuromechanist can you please update the PR merging the https://github.com/bids-standard/bids-specification/pull/2367 and also addressing the conflicts . What is your plan towards making it available for review?

**Comment by neuromechanist** [comment]
Quick update on the recent pushes, now that @yarikoptic's media-files harmonization is merged here:

- The schema caught up with the spec text: the `stim`/`annot` entities are properly wired in, the stim_id checks are rewritten so the validator can actually run them (value patterns now live in the column definitions), and stimulus formats are consistent across the appendix, the file rules, and the docs.
- `/stimuli` now supports **hierarchical subdirectories**. A subdirectory is self-describing, so a stimulus set can move between datasets as a portable unit. A dataset-wide `stimuli.tsv` at the root is RECOMMENDED, even a sparse one listing only the `stimulus_id`s; scoped catalogs amend or override it following the inheritance principle.
- Legacy datasets are untouched: the new rules only kick in when a `stimuli.tsv` exists somewhere under `/stimuli`.
- Everything is tested end to end with a new `beh_stimuli` example (bids-standard/bids-examples#433, now including LLM-generated descriptions and HED annotations for the NSD shared1000 images) and a validator implementation (bids-standard/bids-validator#445); legacy examples like `eeg_matchingpennies` validate unchanged.

File: src/modality-agnostic-files/stimuli.md (hunk 16 above)
Comment (yarikoptic) [reply src/modality-agnostic-files/stimuli.md:34]: the issue still outstanding!

File: src/modality-agnostic-files/stimuli.md [hunk 18]
```diff
@@ -0,0 +1,213 @@
...
+override the corresponding higher-level entries.
+
+For example:
+
+```Text
+stimuli/
+    stimuli.tsv
+    stimuli.json
+    annotations.tsv
+    faces/
```
Comment (yarikoptic) [inline src/modality-agnostic-files/stimuli.md:76]: oh, this is "new" to me concept here, somewhat mimicing the "datatype/" folders we have now (`anat/`, `func/` etc) which sticks out like a sore point... yet to digest this and see how to mediate -- since unlike datatype those are nohow formalized. I feel like we would need some 'cohort-' like entity (similar to `ses-`) to group them and providing `cohorts.tsv` to describe such groups if we are to go this way

File: src/modality-agnostic-files/stimuli.md [hunk 19]
```diff
@@ -0,0 +1,213 @@
...
+        stimuli.tsv          # amends/overrides root entries for stimuli in faces/
```
Comment (yarikoptic) [inline src/modality-agnostic-files/stimuli.md:77]: note that here must be following Inheritance principle but we have not defined it well for .tsv yet!

File: src/modality-agnostic-files/stimuli.md [hunk 20]
```diff
@@ -0,0 +1,213 @@
...
+The definitions of these fields can be found in
+  src/schema/rules/tabular_data/*.yaml
+and a guide for using macros can be found at
+ https://github.com/bids-standard/bids-specification/blob/master/macros_doc.md
+-->
+{{ MACROS___make_columns_table("stimuli.Stimuli") }}
+
+### Example `stimuli.tsv`
+
+```Text
```
Comment (yarikoptic) [inline src/modality-agnostic-files/stimuli.md:149]: There are tsv types
```
src/common-principles.md:```tsv {linenums="1"}
src/common-principles.md-onset      duration        response_time   trial_type      trial_extra
src/common-principles.md-200        20.0    15.8    word    中国人
src/common-principles.md-240        5.0     17.34e-1        visual  n/a
--
src/common-principles.md:```tsvgz {linenums="1"}
src/common-principles.md-200        20.0    15.8    word    中国人
src/common-principles.md-240        5.0     17.34e-1        visual  n/a
```
hence

```suggestion
```tsv
```

File: src/modality-agnostic-files/stimuli.md [hunk 21]
```diff
@@ -0,0 +1,213 @@
...
+
+In cases where the stimulus is not shared, the `stimuli.tsv` file can be used to provide metadata about the stimuli, including the license, copyright, URL, and description. This is similar to the use of `stim-<label>_<suffix>.json` files for individual stimuli files. In the case of conflict between the metadata in the `stimuli.tsv` and `stim-<label>_<suffix>.json` files, the metadata in the `stim-<label>_<suffix>.json` file takes precedence.
+
+## Stimulus Annotations
+
+Annotations of the still images or general description of the stimuli (such as frequency and duration of a beep sound) can be stored in the `stimuli.tsv` as an additional column or `stim-<label>_<suffix>.json` as described above. Here is an example of how annotations can be stored in the `stimuli.tsv` file for an image from the Natural Scene Dataset (NSD):
+
+| stimulus_id   | type  | description                                           | HED                                                                                                                                                               | NSD_id | COCO_id |
```
Comment (yarikoptic) [inline src/modality-agnostic-files/stimuli.md:164]: likely this is intended to be yet another tsv inlined example, not .md table right away, reformat

File: src/modality-agnostic-files/stimuli.md [hunk 22]
```diff
@@ -0,0 +1,213 @@
...
+The definitions of these fields can be found in
+  src/schema/rules/tabular_data/*.yaml
+and a guide for using macros can be found at
+ https://github.com/bids-standard/bids-specification/blob/master/macros_doc.md
+-->
+{{ MACROS___make_columns_table("stimuli.Annotations") }}
+
+### Example `*_annotations.tsv`
+
+```Text
```
Comment (yarikoptic) [inline src/modality-agnostic-files/stimuli.md:185]: also tsv

File: src/modality-agnostic-files/stimuli.md [hunk 23]
```diff
@@ -0,0 +1,213 @@
...
+face01_age   Age group annotation for face01 stimulus
+```
+
+## Referencing Stimulus Identifiers in `events.tsv`
+
+To reference stimulus identifiers in the `events.tsv` file, use the `stim_id` column. The values in the `stim_id` column should represent unique identifiers for the stimuli. Stimulus ID (`stim_id`) should correspond to the unique identifier of the stimulus file in the /stimuli directory and expand to all files (both stimulus and annotation files) that share the same stimulus ID.
+
+Example `events.tsv` file:
+
+| onset | duration | trial_type | response_time | stim_id        |
```
Comment (yarikoptic) [inline src/modality-agnostic-files/stimuli.md:198]: also reformat as tsv inline

**Review by yarikoptic — CHANGES_REQUESTED** [review]
initial, mostly formatting, but raising concern to added groupping folders which are nohow formalized or mostly not inline with the main bids hierarchy building principles (besides hardcoded `datatype/` folders)

note that so far BEP does not introduce the concept of stimuli NEAR the data files as we need for reprostim

File: src/modality-agnostic-files/stimuli.md (hunk 21 above)
Comment (neuromechanist) [reply src/modality-agnostic-files/stimuli.md]: Yes, but it is intentional, the preview is rendering as intended:
image

File: src/modality-agnostic-files/stimuli.md (hunk 23 above)
Comment (neuromechanist) [reply src/modality-agnostic-files/stimuli.md]: Also as above, the rendered markdown looks ok. Mostly cosmetic!
image

Let's discuss during finalization?

**Comment by neuromechanist** [comment]
Thanks @yarikoptic,

Deferring the styling convo to later:
I think we briefly discussed having subdirectories in the stimuli/ dir. Having thousands of images or their annotations, could make subdirs quite useful. I still proposed it as a *convenience* though, meaning that validators and tools flatten them.

On the *TSV inheritance* I agree that it is loosely defined, although I doubt it would be hard to tighten with examples?

File: src/modality-agnostic-files/stimuli.md (hunk 19 above)
Comment (neuromechanist) [reply src/modality-agnostic-files/stimuli.md]: FWIW, `stimuli.tsv` is an extension of `stim-<label>_<suffix>.json`, so the inheritance principle is not far off:

> In cases where the stimulus is not shared, the stimuli.tsv file can be used to provide metadata about the stimuli, including the license, copyright, URL, and description. This is similar to the use of stim-<label>_<suffix>.json files for individual stimuli files. In the case of conflict between the metadata in the stimuli.tsv and stim-<label>_<suffix>.json files, the metadata in the stim-<label>_<suffix>.json file takes precedence.

**Comment by yarikoptic** [comment]
> Deferring the styling convo to later: I think we briefly discussed having subdirectories in the stimuli/ dir. Having thousands of images or their annotations, could make subdirs quite useful. I still proposed it as a _convenience_ though, meaning that validators and tools flatten them.

I do feel that directories could be useful but I just noted that it adds some "inconsistency".  We should also check on what they do in other related BEPs like BEP036 on that

> On the _TSV inheritance_ I agree that it is loosely defined, although I doubt it would be hard to tighten with examples?

I would say we first need to tighten in language since examples can't mandate how things are validated etc, they could just demonstrate.  And overall point here, as .tsv as .json just different containers for metadata, it should be congruent with IP overall as there is now no "overrides" but rather only "summarization" into higher levels... the question is either it is on per-row or per-cell basis -- we better make it clear first in common principles outside of this BEP/examples, and here only demonstrate... what example use cases you had in mind?
